File: java_distance_generator/trunk/src/java/main/com/topcoder/web/tc/distance/data/FlatFileMemberDataAccess.java
/*
 * Copyright (C) 2008 TopCoder Inc., All Rights Reserved.
 */

package com.topcoder.web.tc.distance.data;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import com.topcoder.web.tc.distance.CompetitionType;

/**
 * <p>
 * This class loads XML files named after coder ids (long values), supposed to contain all relevant members, with the
 * member identifying the file first.
 * </p>
 *
 * @author TCSDEVELOPER
 * @version 1.0
 */
public class FlatFileMemberDataAccess implements MemberDataAccess {
    /**
     * <p>
     * Path prepended to the individual file names.
     * </p>
     */
    private String basePath;

    /**
     * <p>
     * Provide a base path used for reading individual files.
     * </p>
     *
     * @param basePath Path to XML files.
     */
    public FlatFileMemberDataAccess(String basePath) {
        this.basePath = basePath;
    }


    /**
     * <p>
     * A string constant indicating the name of the corresponding element in the XML input.
     * </p>
     */
    private static final String CODER_ID_ELEMENT = "coder_id";

    /**
     * <p>
     * A string constant indicating the name of the corresponding element in the XML input.
     * </p>
     */
    private static final String HANDLE_ELEMENT = "handle";

    /**
     * <p>
     * A string constant indicating the suffix of the corresponding element names in the XML input.
     * </p>
     */
    private static final String RATING_SUFFIX = "_rating";

    /**
     * <p>
     * A string constant indicating the name of the corresponding element in the XML input.
     * </p>
     */
    private static final String IMAGE_ELEMENT = "image";

    /**
     * <p>
     * A string constant indicating the name of the corresponding element in the XML input.
     * </p>
     */
    private static final String DISTANCE_ELEMENT = "distance";

    /**
     * <p>
     * A string constant indicating the name of the corresponding element in the XML input.
     * </p>
     */
    private static final String OVERLAP_ELEMENT = "overlap";

    /**
     * <p>
     * A string constant indicating the name of the corresponding element in the XML input.
     * </p>
     */
    private static final String COUNTRY_ELEMENT = "country";

    /**
     * <p>
     * Parses a single coder from XML DOM.
     * </p>
     *
     * @param elem The element to parse.
     * @return The resulting <code>Member</code> object.
     */
    private Member parseCoder(Element elem) {
        NodeList nodes = elem.getChildNodes();
        String countryName = null;
        int geoDistance = -1;
        String handle = null;
        long id = Long.MIN_VALUE;
        String imageName = null;
        int matchOverlap = 0;
        Map<CompetitionType,Integer> ratings = new HashMap<CompetitionType,Integer>();

        for (int i = 0; i < nodes.getLength(); i++) {
            Node node = nodes.item(i);

            if (node instanceof Element) {
                String name = node.getNodeName();
                String value = node.getTextContent();

                if (name.equals(CODER_ID_ELEMENT)) {
                    id = Long.valueOf(value);
                } else if (name.equals(HANDLE_ELEMENT)) {
                    handle = value;
                } else if (name.equals(IMAGE_ELEMENT)) {
                    imageName = value;
                } else if (name.equals(DISTANCE_ELEMENT)) {
                    geoDistance = (int) (Double.parseDouble(value) * 1000.0);
                } else if (name.equals(OVERLAP_ELEMENT)) {
                    matchOverlap = Integer.valueOf(value);
                } else if (name.equals(COUNTRY_ELEMENT)) {
                    countryName = value;
                } else {
                    // Match all competition type names
                    for (CompetitionType type : CompetitionType.values()) {
                        if (name.equals(type.toString().toLowerCase() + RATING_SUFFIX)) {
                            ratings.put(type, Integer.valueOf(value));
                        }
                    }
                }
            }
        }

        return new Member(id, handle, ratings, countryName, imageName, geoDistance, matchOverlap);
    }

    /**
     * <p>
     * Gets all member objects from the XML file for the matching id.
     * </p>
     *
     * @param id The id to read from.
     * @return The final list.
     */
    private List<Member> getMembers(long id) throws MemberDataAccessException {
        try {
            File file = new File(basePath + Long.toString(id) + ".xml");
            List<Member> result = new ArrayList<Member>();

            if (!file.exists()) {
                return result;
            }

            DocumentBuilder builder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
            Document document = builder.parse(file);

            // root element map
            NodeList nodes = document.getChildNodes();
            for (int i = 0; i < nodes.getLength(); i++) {
                Node node = nodes.item(i);
                if (node instanceof Element) {
                    NodeList nodes2 = node.getChildNodes();
                    // coder elements
                    for (int j = 0; j < nodes2.getLength(); j++) {
                        Node node2 = nodes2.item(j);
                        if (node2 instanceof Element) {
                            result.add(parseCoder((Element) node2));
                        }
                    }
                }
            }

            return result;
        } catch (IOException e) {
            throw new MemberDataAccessException("Error reading coder file.", e);
        } catch (SAXException e) {
            throw new MemberDataAccessException("Error parsing coder file.", e);
        } catch (ParserConfigurationException e) {
            throw new MemberDataAccessException("Error parsing configuration when reading coder file.", e);
        }
    }

    /**
     * <p>
     * Loads the single, first, member in the XML file for the specified coder.
     * </p>
     *
     * @param id The id of the coder.
     * @return A <code>Member</code> object for the coder, or null if there is no file for this coder.
     *
     * @throws MemberDataAccessException Thrown for any parsing error.
     */
    public Member getMember(long id) throws MemberDataAccessException {
        List<Member> members = getMembers(id);
        if (members.size() == 0) {
            return null;
        }

        return members.get(0);
    }

    /**
     * <p>
     * Loads all related members (not the "origin member") from the file of the specified coder.
     * </p>
     *
     * @param id The id of the coder.
     * @param competitionTypes The competition types requested.
     *
     * @return A <code>List</code> of <code>Member</code> objects, possibly empty.
     */
    public List<Member> getRelatedMembers(long id,
            EnumSet<CompetitionType> competitionTypes)
            throws MemberDataAccessException {
        List<Member> result = new ArrayList<Member>();

        List<Member> members = getMembers(id);
        if (members.size() >= 2) {
            result.addAll(members.subList(1, members.size()));
        }

        return result;
    }
}

File: dotnet_distance_generator/trunk/src/csharp/main/TopCoder/Web/Distance/Data/FlatFileMemberDataAccess.cs
/*
 * Copyright (c) 2008, TopCoder, Inc. All rights reserved.
 */

using System;
using System.Collections.Generic;
using System.Text;

using System.Xml;
using System.IO;

namespace TopCoder.Web.Distance.Data
{

    /// <summary>
    /// This class loads XML files named after coder ids (long values),
    /// supposed to contain all relevant members, with the
    /// member identifying the file first.
    /// </summary>
    /// <author>TCSDESIGNER</author>
    /// <author>hotblue</author>
    /// <copyright>Copyright (c) 2008, TopCoder, Inc. All rights reserved.</copyright>
    /// <version>1.0</version>
    public class FlatFileMemberDataAccess : IMemberDataAccess
    {
        /// <summary>
        /// The base path.
        /// </summary>
        private string basePath;

        /// <summary>
        /// Provide a base path used for reading individual files.
        /// </summary>
        /// <param name="basePath">The path to the coder data.</param>
        public FlatFileMemberDataAccess(string basePath)
        {
            this.basePath = basePath;
        }

        /// <summary>
        /// A string constant indicating the name of the corresponding element in the XML input.
        /// </summary>
        private const string CODER_ID_ELEMENT = "coder_id";

        /// <summary>
        /// A string constant indicating the name of the corresponding element in the XML input.
        /// </summary>
        private const string HANDLE_ELEMENT = "handle";

        /// <summary>
        /// A string constant indicating the suffix of the corresponding element names in the XML input.
        /// </summary>
        private const string RATING_SUFFIX = "_rating";

        /// <summary>
        /// A string constant indicating the name of the corresponding element in the XML input.
        /// </summary>
        private const string IMAGE_ELEMENT = "image";

        /// <summary>
        /// A string constant indicating the name of the corresponding element in the XML input.
        /// </summary>
        private const string DISTANCE_ELEMENT = "distance";

        /// <summary>
        /// A string constant indicating the name of the corresponding element in the XML input.
        /// </summary>
        private const string OVERLAP_ELEMENT = "overlap";

        /// <summary>
        /// A string constant indicating the name of the corresponding element in the XML input.
        /// </summary>
        private const string COUNTRY_ELEMENT = "country";

        /// <summary>
        /// Parses a single coder from XML DOM.
        /// </summary>
        /// <param name="coder">The element to parse.</param>
        /// <returns>The resulting <see cref="Member"/> object.</returns>
        private Member ParseCoder(XmlNode coder)
        {
            String countryName = null;
            int geoDistance = -1;
            String handle = null;
            long id = long.MinValue;
            String imageName = null;
            int matchOverlap = 0;
            Dictionary<CompetitionTypes, int> ratings = new Dictionary<CompetitionTypes, int>();

            foreach (XmlNode subNode in coder)
            {
                String name = subNode.Name;
                String value = subNode.InnerText;

                if (name == CODER_ID_ELEMENT)
                {
                    id = long.Parse(value);
                }
                else if (name == HANDLE_ELEMENT)
                {
                    handle = value;
                }
                else if (name == IMAGE_ELEMENT)
                {
                    imageName = value;
                }
                else if (name == DISTANCE_ELEMENT)
                {
                    geoDistance = (int)(double.Parse(value) * 1000);
                }
                else if (name == OVERLAP_ELEMENT)
                {
                    matchOverlap = int.Parse(value);
                }
                else if (name == COUNTRY_ELEMENT)
                {
                    countryName = value;
                }
                else
                {
                    // Match all competition type names
                    // for (CompetitionType type : CompetitionType.values())
                    foreach (CompetitionTypes type in Enum.GetValues(typeof(CompetitionTypes)))
                    {
                        if (name == type.ToString().ToLower() + RATING_SUFFIX)
                        {
                            ratings[type] = int.Parse(value);
                        }
                    }
                }
            }

            foreach (CompetitionTypes t in Enum.GetValues(typeof(CompetitionTypes)))
            {
                if (!ratings.ContainsKey(t))
                {
                    ratings[t] = 0;
                }
            }

            return new Member(id, handle, ratings, countryName, imageName, geoDistance, matchOverlap);
        }

        /// <summary>
        /// Gets all member objects from the XML file for the matching id.
        /// </summary>
        /// <param name="id">The id to read from.</param>
        /// <returns>The final list.</returns>
        /// <exception cref="MemberDataAccessException">If error occurs during retrieval.</exception>
        private IList<Member> GetMembers(long id)
        {
            try
            {
                XmlDocument xmld = new XmlDocument();
                xmld.Load(basePath + id.ToString() + ".xml");
                List<Member> result = new List<Member>();

                if (xmld == null)
                {
                    return result;
                }

                XmlNode root = xmld.ChildNodes[0];

                // root element map
                foreach (XmlNode coder in root.ChildNodes)
                {
                    // coder elements
                    result.Add(ParseCoder(coder));
                }

                return result;
            }
            catch (IOException e)
            {
                throw new MemberDataAccessException("Error reading coder file.", e);
            }
            catch (XmlException e)
            {
                throw new MemberDataAccessException("Error parsing coder file.", e);
            }
        }

        /// <summary>
        /// Loads the single, first, member in the XML file for the specified coder.
        /// b</summary>
        /// <param name="id">The id of the coder.</param>
        /// <returns>
        /// A Member object for the coder, or null if there is no file for this coder.
        /// </returns>
        /// <exception cref="MemberDataAccessException">
        /// Thrown for any parsing error.
        /// </exception>
        public Member GetMember(long id)
        {
            IList<Member> members = GetMembers(id);
            if (members.Count == 0)
            {
                return null;
            }

            return members[0];
        }

        /// <summary>
        /// Loads all related members (not the "origin member") from the file of
        /// the specified coder.
        /// </summary>
        /// <param name="id">
        /// The id of the coder.
        /// </param>
        /// <param name="competitionTypes">
        /// The competition types requested.
        /// </param>
        /// <returns>
        /// A list containing all the members related to the origin member.
        /// </returns>
        public IList<Member> GetRelatedMembers(long id, CompetitionTypes competitionTypes)
        {
            List<Member> result = new List<Member>();
            IList<Member> members = GetMembers(id);

            if (members.Count >= 2)
            {
                members.Remove(members[0]);
                result = new List<Member>(members);
            }

            return result;
        }
    }

}
